src/js/common/parser.js

```diff
@@ -0,0 +1,281 @@
+    return pointerSegments.concat(childPath).join("/");
+};
+
+/**
+ *
+ * Static function to determine the JSON pointer to an error definition given the JSON pointer to the failure returned
+ * by AJV. We will begin with something like `#/field1/type` and return something like `#/field1/errors`.
+ *
+ * @param failurePointer {String} - A JSON pointer representing the failing rule as reporting by AJV.
+ * @returns An {Object} representing the `errors` block for the given field.
+ *
+ */
+gpii.schema.parser.getFieldErrorsFromFailure = function (failurePointer) {
+    var parentJsonPointer = gpii.schema.parser.getParentJsonPointer(failurePointer);
+    return gpii.schema.parser.getChildJsonPointer(parentJsonPointer, "errors");
+};
+
+
+/**
+ *
+ * Static function to determine the JSON pointer that points to a missing required field.  Since required fields are
+ * defined in arrays within the enclosing object, these will be references like `#/required/1`.
+ *
+ * @param that - The parser component itself.
+ * @param schemaKey {String} - The filename/id of the schema we are working with.
+ * @param failurePointer {String} - A JSON pointer representing the validation rule that was broken.
+ * @param propertyToMatch {String} - The missing property as reported by AJV.
+ * @returns A JSON pointer {String} that can be used to look up the relevant error data from the schema.
+ *
+ */
+gpii.schema.parser.getRequiredFieldPointer = function (that, schemaKey, failurePointer, propertyToMatch) {
+    var requireDefinitions = that.resolveJsonPointer(schemaKey, failurePointer);
+    var requirementIndex = fluid.find(requireDefinitions, function (value, index) {
+        if (value === propertyToMatch) { return index; }
+    });
+
+    // The path to the message is relative to the parent
+    if (requirementIndex !== undefined && requirementIndex !== null) {
+        var parentPointer = gpii.schema.parser.getParentJsonPointer(failurePointer);
+        var errorsPointer = gpii.schema.parser.getChildJsonPointer(parentPointer, "required");
+        return gpii.schema.parser.getChildJsonPointer(errorsPointer, requirementIndex);
+    }
+    else {
+        return undefined;
+    }
+};
+
+/**
+ *
+ * A static function to return the last segment of a given JSON pointer.
+ *
+ * @param jsonPointer {String} - The original JSON pointer.
+ * @returns A {String} representing the last segment of the original pointer.
+ */
+gpii.schema.parser.getLastJsonPointerSegment = function (jsonPointer) {
+    var segments = jsonPointer.split("/");
+    return segments[segments.length - 1];
+};
+
+/**
+ *
+ * A model listener to dereference and cache a dereferenced version of schemas as they are added to `model.schemas`.
+ */
+gpii.schema.parser.updateSchemas = function (that) {
+    var promises = [];
+    fluid.each(
+        that.model.schemas, function (schemaContent, schemaKey) {
+        if (!that.dereferencedSchemas[schemaKey]) {
```

Comment:
I think there is a bug here, in that "dereferencedSchemas" is never cleared, which implies that it can only ever be updated once. You should add a fix and test case for the situation where a schema is fetched once and then updated.